File: missing-item-benchmark/c/missing-item.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>
#include <string.h>

// Function to measure execution time
double time_function(int (*func)(int*, int, int*, int), int* a, int size_a, int* b, int size_b) {
    clock_t start = clock();
    func(a, size_a, b, size_b);  // We don't need the result for timing
    clock_t end = clock();
    double elapsed = ((double)(end - start)) / CLOCKS_PER_SEC;
    return elapsed;
}

// XOR method - original approach
int find_missing_xor_original(int* a, int size_a, int* b, int size_b) {
    int xor_a = 0;
    int xor_b = 0;
    
    for (int i = 0; i < size_a; i++) {
        xor_a ^= a[i];
    }
    
    for (int i = 0; i < size_b; i++) {
        xor_b ^= b[i];
    }
    
    return xor_a ^ xor_b;
}

// XOR method - optimized single loop
int find_missing_xor_optimized(int* a, int size_a, int* b, int size_b) {
    int result = 0;
    
    for (int i = 0; i < size_a; i++) {
        result ^= a[i];
    }
    
    for (int i = 0; i < size_b; i++) {
        result ^= b[i];
    }
    
    return result;
}

// Sum difference method
int find_missing_sum(int* a, int size_a, int* b, int size_b) {
    long long sum_a = 0;
    long long sum_b = 0;
    
    for (int i = 0; i < size_a; i++) {
        sum_a += a[i];
    }
    
    for (int i = 0; i < size_b; i++) {
        sum_b += b[i];
    }
    
    return (int)(sum_a - sum_b);
}

// Linear search method (naive approach)
int find_missing_linear(int* a, int size_a, int* b, int size_b) {
    for (int i = 0; i < size_a; i++) {
        int count_a = 0, count_b = 0;
        
        // Count occurrences in array a
        for (int j = 0; j < size_a; j++) {
            if (a[j] == a[i]) count_a++;
        }
        
        // Count occurrences in array b
        for (int j = 0; j < size_b; j++) {
            if (b[j] == a[i]) count_b++;
        }
        
        if (count_a != count_b) {
            return a[i];
        }
    }
    return -1; // Not found
}

// Utility function to copy array
void copy_array(int* src, int* dest, int size) {
    for (int i = 0; i < size; i++) {
        dest[i] = src[i];
    }
}

// Utility function to remove random element
int remove_random_element(int* array, int size) {
    int index = rand() % size;
    int removed = array[index];
    
    // Shift elements to remove the element at index
    for (int i = index; i < size - 1; i++) {
        array[i] = array[i + 1];
    }
    
    return removed;
}

int main() {
    srand(time(NULL));
    
    printf("XOR Benchmark - C Implementation\n");
    printf("=================================\n\n");
    
    // Test different sizes (powers of 2)
    int sizes[] = {2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768};
    int num_sizes = sizeof(sizes) / sizeof(sizes[0]);
    
    // Function pointers for different algorithms
    int (*algorithms[])(int*, int, int*, int) = {
        find_missing_xor_original,
        find_missing_xor_optimized,
        find_missing_sum,
        find_missing_linear
    };
    
    const char* algorithm_names[] = {
        "XOR (original)",
        "XOR (optimized)", 
        "Sum difference",
        "Linear search"
    };
    
    int num_algorithms = sizeof(algorithms) / sizeof(algorithms[0]);
    
    for (int s = 0; s < num_sizes; s++) {
        int size = sizes[s];
        printf("=== Testing with array size: %d ===\n", size);
        
        // Create array with random integers
        int* a = malloc(size * sizeof(int));
        int* b = malloc((size - 1) * sizeof(int));
        
        for (int i = 0; i < size; i++) {
            a[i] = rand() % size;
        }
        
        // Copy array a to b, then remove one element
        copy_array(a, b, size);
        remove_random_element(b, size);
        int size_b = size - 1;
        
        // Run each algorithm multiple times and calculate average
        const int iterations = 10;
        double avg_times[num_algorithms];
        int results[num_algorithms];
        
        for (int alg = 0; alg < num_algorithms; alg++) {
            double total_time = 0.0;
            
            for (int iter = 0; iter < iterations; iter++) {
                double elapsed = time_function(algorithms[alg], a, size, b, size_b);
                total_time += elapsed;
            }
            
            avg_times[alg] = total_time / iterations;
            results[alg] = algorithms[alg](a, size, b, size_b);
            
            printf("Average time using %s: %.8f seconds\n", 
                   algorithm_names[alg], avg_times[alg]);
        }
        
        // Find fastest method
        int fastest_idx = 0;
        for (int i = 1; i < num_algorithms; i++) {
            if (avg_times[i] < avg_times[fastest_idx]) {
                fastest_idx = i;
            }
        }
        
        printf("Fastest method: %s (%.8f seconds)\n", 
               algorithm_names[fastest_idx], avg_times[fastest_idx]);
        
        // Check for result consistency
        int consistent = 1;
        for (int i = 1; i < num_algorithms; i++) {
            if (results[i] != results[0]) {
                consistent = 0;
                break;
            }
        }
        
        if (consistent) {
            printf("All methods returned the same result: %d\n", results[0]);
        } else {
            printf("Discrepancy found in results:\n");
            for (int i = 0; i < num_algorithms; i++) {
                printf("  %s: %d\n", algorithm_names[i], results[i]);
            }
        }
        
        printf("\n");
        
        free(a);
        free(b);
    }
    
    return 0;
}

File: missing-item-benchmark/go/missing-item.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

// timeFunction measures the execution time of a function
func timeFunction(fn func([]int, []int) int, a, b []int) (int, time.Duration) {
	start := time.Now()
	result := fn(a, b)
	elapsed := time.Since(start)
	return result, elapsed
}

// findMissingXOROriginal - XOR method using separate loops
func findMissingXOROriginal(a, b []int) int {
	xorA := 0
	xorB := 0

	for _, num := range a {
		xorA ^= num
	}

	for _, num := range b {
		xorB ^= num
	}

	return xorA ^ xorB
}

// findMissingXOROptimized - XOR method using single result variable
func findMissingXOROptimized(a, b []int) int {
	result := 0

	for _, num := range a {
		result ^= num
	}

	for _, num := range b {
		result ^= num
	}

	return result
}

// findMissingSum - Sum difference method
func findMissingSum(a, b []int) int {
	sumA := 0
	sumB := 0

	for _, num := range a {
		sumA += num
	}

	for _, num := range b {
		sumB += num
	}

	return sumA - sumB
}

// findMissingLinear - Linear search method using map
func findMissingLinear(a, b []int) int {
	// Count occurrences in both slices
	countA := make(map[int]int)
	countB := make(map[int]int)

	for _, num := range a {
		countA[num]++
	}

	for _, num := range b {
		countB[num]++
	}

	// Find the element with different counts
	for num, countInA := range countA {
		if countB[num] != countInA {
			return num
		}
	}

	return 0 // Should never reach here if input is valid
}

// findMissingSet - Set difference method using map lookup
func findMissingSet(a, b []int) int {
	// Create a set from b
	setB := make(map[int]int)
	for _, num := range b {
		setB[num]++
	}

	// Check each element in a
	countA := make(map[int]int)
	for _, num := range a {
		countA[num]++
	}

	for num, countInA := range countA {
		if setB[num] < countInA {
			return num
		}
	}

	return 0 // Should never reach here if input is valid
}

// generateTestData creates test arrays with one missing element
func generateTestData(size int) ([]int, []int) {
	// Create array a with random integers
	a := make([]int, size)
	for i := 0; i < size; i++ {
		a[i] = rand.Intn(size)
	}

	// Create array b as copy of a, then remove one random element
	b := make([]int, len(a))
	copy(b, a)

	// Remove a random element
	removeIndex := rand.Intn(len(b))
	b = append(b[:removeIndex], b[removeIndex+1:]...)

	return a, b
}

func main() {
	// Test with different list sizes
	sizes := []int{2, 4, 8, 16, 32, 64, 128, 256, 512, 1024, 2048, 4096, 8192, 16384, 32768}

	// Define test functions
	testFunctions := map[string]func([]int, []int) int{
		"xor_original":  findMissingXOROriginal,
		"xor_optimized": findMissingXOROptimized,
		"sum":           findMissingSum,
		"linear":        findMissingLinear,
		"set":           findMissingSet,
	}

	// Seed random number generator
	rand.Seed(time.Now().UnixNano())

	for _, size := range sizes {
		fmt.Printf("\n=== Testing with list size: %d ===\n", size)

		// Generate test data
		a, b := generateTestData(size)

		// Run multiple times and average
		timeResults := make(map[string][]time.Duration)
		for key := range testFunctions {
			timeResults[key] = make([]time.Duration, 0, 10)
		}

		var results []int

		// Run 10 times for better average
		for i := 0; i < 10; i++ {
			for key, fn := range testFunctions {
				result, elapsed := timeFunction(fn, a, b)
				timeResults[key] = append(timeResults[key], elapsed)
				if i == 0 { // Only store results from first run for comparison
					results = append(results, result)
				}
			}
		}

		// Calculate averages
		avgTimes := make(map[string]time.Duration)
		for key, times := range timeResults {
			var total time.Duration
			for _, t := range times {
				total += t
			}
			avgTimes[key] = total / time.Duration(len(times))
		}

		// Print results
		fmt.Printf("Average time using XOR (original): %v\n", avgTimes["xor_original"])
		fmt.Printf("Average time using XOR (optimized): %v\n", avgTimes["xor_optimized"])
		fmt.Printf("Average time using Sum: %v\n", avgTimes["sum"])
		fmt.Printf("Average time using Linear: %v\n", avgTimes["linear"])
		fmt.Printf("Average time using Set: %v\n", avgTimes["set"])

		// Find the fastest method
		var fastestMethod string
		var fastestTime time.Duration = time.Hour // Start with a very large time

		times := map[string]time.Duration{
			"XOR (original)":  avgTimes["xor_original"],
			"XOR (optimized)": avgTimes["xor_optimized"],
			"Sum":             avgTimes["sum"],
			"Linear":          avgTimes["linear"],
			"Set":             avgTimes["set"],
		}

		for method, t := range times {
			if t < fastestTime {
				fastestTime = t
				fastestMethod = method
			}
		}

		fmt.Printf("Fastest method: %s (%v)\n", fastestMethod, fastestTime)

		// Check for discrepancies - all results should be the same
		firstResult := results[0]
		allSame := true
		for _, result := range results[1:] {
			if result != firstResult {
				allSame = false
				break
			}
		}

		if !allSame {
			fmt.Println("Discrepancy found in results:")
			i := 0
			for key := range testFunctions {
				fmt.Printf("%s: %d\n", key, results[i])
				i++
			}
		} else {
			fmt.Printf("All methods returned the same result: %d\n", firstResult)
		}
	}
}
